engine.go

```diff
@@ -39,125 +46,136 @@ type Finding struct {
-	if err != nil {
-		return err
-	}
-	return writeInOutputFile(outputFile, bytesToWrite)
+// Engine contains all the engine necessary data
+type Engine struct {
+	poolSize   int
+	extensions []string
 }
 
-func RunMaxUnitsByAnalysis(document []Unit, rules []Rule, maxUnitPerAnalysis int) (documentFindings []Finding) {
-	listDocuments := breakTextUnitsIntoLimitOfUnit(document, maxUnitPerAnalysis)
-	for key, units := range listDocuments {
-		logger.LogDebugWithLevel(fmt.Sprintf("Start run analysis %v/%v", key, len(listDocuments)), loggerEnums.DebugLevel)
-		documentFindings = append(documentFindings, Run(units, rules)...)
+// NewEngine creates a new engine instance with all necessary data.
+// extensions argument represents which extension the engine should apply the rules
+// poolSize represents the number of go routines to open (Default is 10)
+func NewEngine(poolSize int, extensions ...string) *Engine {
+	return &Engine{
+		poolSize:   poolSize,
+		extensions: extensions,
 	}
-	return documentFindings
 }
 
-func breakTextUnitsIntoLimitOfUnit(allUnits []Unit, maxUnitsPerAnalysis int) (units [][]Unit) {
-	units = [][]Unit{}
-	startIndex := 0
-	endIndex := maxUnitsPerAnalysis
-	for i := 0; i < getTotalTextUnitsToRunByAnalysis(allUnits, maxUnitsPerAnalysis); i++ {
-		units = append(units, []Unit{})
-		units = toBreakUnitsAddUnitAndUpdateStartEndIndex(allUnits, units, startIndex, endIndex, i)
-		startIndex = endIndex + 1
-		endIndex += maxUnitsPerAnalysis
-	}
-	return units
-}
+// Run walks through projectPath and runs the method Rule.Run in a pool of goroutines
+// if an error is found when executes Rule.Run method it cancels current running go routines and return
+// valid findings and the error
+// nolint:funlen // necessary complexity, breaking this function will lead to a really complex code
+func (e *Engine) Run(ctx context.Context, projectPath string, rules ...Rule) ([]Finding, error) {
+	var findings SafeFindings
 
-func getTotalTextUnitsToRunByAnalysis(textUnits []Unit, maxUnitsPerAnalysis int) int {
-	totalTextUnits := len(textUnits)
-	if totalTextUnits <= maxUnitsPerAnalysis {
-		return 1
+	paths, err := e.getValidFilePaths(projectPath)
+	if err != nil {
+		return nil, err
 	}
-	totalUnitsToRun := float64(totalTextUnits / maxUnitsPerAnalysis)
-	// nolint:staticcheck // is necessary usage pointless in math.ceil
-	return int(math.Ceil(totalUnitsToRun))
-}
 
-func toBreakUnitsAddUnitAndUpdateStartEndIndex(
-	allUnits []Unit, unitsToAppend [][]Unit, startIndex, endIndex, i int) [][]Unit {
-	if len(allUnits[startIndex:]) <= endIndex {
-		for k := range allUnits[startIndex:] {
-			unitsToAppend[i] = append(unitsToAppend[i], allUnits[k])
-		}
-	} else {
-		for k := range allUnits[startIndex:endIndex] {
-			unitsToAppend[i] = append(unitsToAppend[i], allUnits[k])
+	wg := sync.WaitGroup{}
+
+	workerPool := pool.Default(e.poolSize)
+	defer workerPool.Release()
+
+	group, _ := errgroup.WithContext(ctx)
+
+	for _, path := range paths {
+		wg.Add(1)
```

Comment:
We put this wg.Add before the loop using `wg.Add(len(paths))`

Reply:
Done!